{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c135d995",
   "metadata": {},
   "source": [
    "# version update colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5677a6d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import depencies\n",
    "from scipy.sparse import csr_matrix\n",
    "import tables\n",
    "import h5py\n",
    "import hdf5plugin\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.colors as colors\n",
    "import numpy as np\n",
    "import seaborn as sb\n",
    "from scipy import signal\n",
    "import glob\n",
    "from scipy.signal import savgol_filter\n",
    "from scipy.ndimage import gaussian_filter1d as gfilter\n",
    "import pandas as pd\n",
    "\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "import matplotlib as mpl\n",
    "\n",
    "# scale figure\n",
    "mpl.rcParams['figure.dpi'] = 200"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cae39862-4c3b-4cb2-846a-5dd765c91ffe",
   "metadata": {},
   "source": [
    "# Import contact frequency matrices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00ee05df",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import find raw data files\n",
    "\n",
    "file_names = []\n",
    "for name in glob.glob('./Raw data/*_5kb.h5'):\n",
    "    file_names.append(name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef3274c0-a8db-4660-9bd9-f9dc8afdc883",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_names"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "feb439e6",
   "metadata": {},
   "source": [
    "# Define functions useful for analysis\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f18fcb53",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import HiC heat map data as a matrix\n",
    "\n",
    "#argument:\n",
    "# filename: input name of the file you wish to import matrix for\n",
    "\n",
    "#output:\n",
    "#the heat map itself as a matrix\n",
    "def make_matrix(filename):\n",
    "    with tables.open_file(filename, 'r') as f:\n",
    "        parts = {}\n",
    "        try:\n",
    "            for matrix_part in ('data', 'indices', 'indptr', 'shape'):\n",
    "                parts[matrix_part] = getattr(f.root.matrix, matrix_part).read()\n",
    "        except Exception as e:\n",
    "            log.info('No h5 file. Please check parameters concerning the file type!')\n",
    "            e\n",
    "        matrix = csr_matrix(tuple([parts['data'], parts['indices'], parts['indptr']]),\n",
    "                            shape=parts['shape'])\n",
    "        matrix_array = matrix.toarray()\n",
    "    return matrix_array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4c91791",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import interval information, contains list of indices and chromosome that they correspond to\n",
    "\n",
    "#argument:\n",
    "# filename: input name of the file you wish to import intervals for\n",
    "\n",
    "#output:\n",
    "#interval_list: list of chromosomes, indices\n",
    "#keychr: keys to access different list from interval_list\n",
    "\n",
    "def get_intervals(filename):\n",
    "    file = h5py.File(filename)\n",
    "    key_list = file['intervals'].keys()\n",
    "    interval_list = {}\n",
    "    for key in key_list:\n",
    "        interval_list[key] = file['intervals'][key][()]\n",
    "    keychr = list(key_list)\n",
    "    return interval_list, keychr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd0b7326",
   "metadata": {},
   "outputs": [],
   "source": [
    "#generate a matrix for specific chromosome vs chromosome interaction\n",
    "\n",
    "#argument:\n",
    "#ma: matrix of heatmap\n",
    "#chromosome_x: chromosome to use for x axis\n",
    "#chromosome_y: chromosome to use for y axis\n",
    "\n",
    "#output:\n",
    "#matrix of desired region\n",
    "\n",
    "def call_region(ma,chromosome_x,chromosome_y):\n",
    "    start_x = indices[chromosome_x]['start']\n",
    "    end_x = indices[chromosome_x]['end']\n",
    "    \n",
    "    start_y = indices[chromosome_y]['start']\n",
    "    end_y = indices[chromosome_y]['end']\n",
    "    \n",
    "    return  ma[start_x:end_x,start_y:end_y]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e08ed4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chunky code to get the end index for regions, requried for get_indices function\n",
    "# argument:\n",
    "# chrlist: full list of which chromosome each index correspond to\n",
    "\n",
    "#output:\n",
    "#end: end index for each chromosome\n",
    "\n",
    "def get_end_idx(chrlist):\n",
    "    start = {}\n",
    "    end = {}\n",
    "    i = 0\n",
    "    for chromes in chrlist:\n",
    "        if chromes == b'chrIV':\n",
    "            end[chromes] = i\n",
    "            i+=1\n",
    "        elif chromes == b'chrXV':\n",
    "            end[chromes] = i\n",
    "            i+=1\n",
    "        elif chromes == b'chrVII':\n",
    "            end[chromes] = i\n",
    "            i+=1\n",
    "        elif chromes == b'chrXII':\n",
    "            end[chromes] = i\n",
    "            i+=1\n",
    "        elif chromes == b'chrXVI':\n",
    "            end[chromes] = i\n",
    "            i+=1\n",
    "        elif chromes == b'chrXIII':\n",
    "            end[chromes] = i\n",
    "            i+=1\n",
    "        elif chromes == b'chrII':\n",
    "            end[chromes] = i\n",
    "            i+=1\n",
    "        elif chromes == b'chrXIV':\n",
    "            end[chromes] = i\n",
    "            i+=1\n",
    "        elif chromes == b'chrX':\n",
    "            end[chromes] = i\n",
    "            i+=1\n",
    "        elif chromes == b'chrXI':\n",
    "            end[chromes] = i\n",
    "            i+=1\n",
    "        elif chromes == b'chrV':\n",
    "            end[chromes] = i\n",
    "            i+=1\n",
    "        elif chromes == b'chrVIII':\n",
    "            end[chromes] = i\n",
    "            i+=1\n",
    "        elif chromes == b'chrIX':\n",
    "            end[chromes] = i\n",
    "            i+=1\n",
    "        elif chromes == b'chrIII':\n",
    "            end[chromes] = i\n",
    "            i+=1\n",
    "        elif chromes == b'chrVI':\n",
    "            end[chromes] = i\n",
    "            i+=1\n",
    "        elif chromes == b'chrI':\n",
    "            end[chromes] = i\n",
    "            i+=1\n",
    "        elif chromes == b'chrM':\n",
    "            end[chromes] = i\n",
    "            i+=1\n",
    "        else:\n",
    "            break\n",
    "    return end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0459035",
   "metadata": {},
   "outputs": [],
   "source": [
    "# makes a dictionary of each chromosomes start and end index\n",
    "\n",
    "# argument:\n",
    "# chromosomes: the list of chromosomes\n",
    "# end_idx: output of get_end_idx\n",
    "\n",
    "#output:\n",
    "#idxs: dictionary of each chromosomes start and end index\n",
    "\n",
    "def get_indices(chromosomes,end_idx):\n",
    "    idxs = {}\n",
    "    next_start = 0\n",
    "    for chromes in chromosomes:\n",
    "        idxs[chromes] = {}\n",
    "        idxs[chromes]['end'] = end_idx[chromes]+1\n",
    "\n",
    "        if chromes == str(b'chrIV'):\n",
    "            idxs[chromes]['start'] = 0\n",
    "        else:\n",
    "            idxs[chromes]['start'] = next_start\n",
    "        next_start = end_idx[chromes]+1\n",
    "        \n",
    "    return idxs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec4a0d2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#divide 2 matrices\n",
    "\n",
    "#argument: \n",
    "#nom_matrix: matrix on the nominator\n",
    "#denom_matrix: matrix on the denominator\n",
    "#log2: whether to apply log2 to output matrix\n",
    "\n",
    "#output:\n",
    "#matrix: matrix after division\n",
    "\n",
    "def divide_matrices(nom_matrix, denom_matrix, log2 = True):\n",
    "    divide_matrix = nom_matrix/denom_matrix\n",
    "#     divide_matrix[~np.isfinite(divide_matrix)] = 0\n",
    "    if log2:\n",
    "        matrix = np.log2(divide_matrix)\n",
    "        matrix[~np.isfinite(matrix)] = 0\n",
    "    else:\n",
    "        matrix = divide_matrix\n",
    "        matrix[~np.isfinite(matrix)] = 0\n",
    "    return matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65a3bbc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#make the triangular matrix a full matrix by duplicating its diagonal counterpart\n",
    "\n",
    "#argument:\n",
    "#ma: matrix of heatmap\n",
    "\n",
    "#output:\n",
    "# full_matrix\n",
    "\n",
    "def make_symmetric(ma):\n",
    "    rotmatrix = np.transpose(ma)\n",
    "    np.fill_diagonal(rotmatrix,0)\n",
    "    full_matrix = ma + rotmatrix\n",
    "    return full_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c3f2bd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import h5 files and produce the full matrix\n",
    "\n",
    "#argument:\n",
    "#filename: file name of the data to import\n",
    "\n",
    "#output:\n",
    "# full_matrix\n",
    "\n",
    "def make_symm_matrix(filename):\n",
    "    with tables.open_file(filename, 'r') as f:\n",
    "        parts = {}\n",
    "        try:\n",
    "            for matrix_part in ('data', 'indices', 'indptr', 'shape'):\n",
    "                parts[matrix_part] = getattr(f.root.matrix, matrix_part).read()\n",
    "        except Exception as e:\n",
    "            log.info('No h5 file. Please check parameters concerning the file type!')\n",
    "            e\n",
    "        matrix = csr_matrix(tuple([parts['data'], parts['indices'], parts['indptr']]),\n",
    "                            shape=parts['shape'])\n",
    "        matrix_array = matrix.toarray()\n",
    "        \n",
    "        full_ma = make_symmetric(matrix_array)\n",
    "\n",
    "    return full_ma"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60fb3d23",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2d filter the matrix, eliminate high frequency noise\n",
    "\n",
    "#argument:\n",
    "#ma: matrix\n",
    "#filter_length: length to apply the filter, larger length decreases max signal frequency\n",
    "\n",
    "#output:\n",
    "#filtered_ma: matrix after apply filtering\n",
    "\n",
    "def filtering(ma, filter_length = 5):\n",
    "    filtered_ma = signal.spline_filter(ma,lmbda=filter_length)\n",
    "    return filtered_ma"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a41d8adb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot the matrix as a heatmap\n",
    "\n",
    "def plot_matrix(ma, filt = None,filter_len=0, symmetric = True, hlim = None,vlim =None, center=False, interpolation =None, cmap = 'RdYlBu_r', fig_size = 15,vcenter=None, vmin = None, vmax = None ):\n",
    "    fig, ax = plt.subplots()\n",
    "\n",
    "    ma[~np.isfinite(ma)] = 0\n",
    "    if filt:\n",
    "        filtered_ma = filtering(ma,filter_length = filter_len)\n",
    "    else:\n",
    "        filtered_ma = ma\n",
    "    if symmetric:\n",
    "        full_ma = filtered_ma\n",
    "    else:\n",
    "        full_ma = make_symmetric(filtered_ma)\n",
    "    if center:\n",
    "        norm = colors.TwoSlopeNorm(vmin=vmin, vcenter=vcenter, vmax=vmax)\n",
    "        im = ax.imshow(full_ma,cmap = cmap,norm=norm)\n",
    "    else:\n",
    "        im = ax.imshow(full_ma,cmap = cmap,interpolation = interpolation,vmin = vmin, vmax = vmax)\n",
    "    plt.ylim(vlim)\n",
    "    plt.xlim(hlim)\n",
    "    cb = fig.colorbar(im,fraction=0.046, pad=0.04)\n",
    "#     cb.set_ticks([0,1,2,3,4])\n",
    "#     cb.set_ticklabels(['$10^0$','$10^1$','$10^2$','$10^3$','$10^4$'])\n",
    "#     plt.xlabel('sites /5kB')\n",
    "#     plt.ylabel('sites /5kB')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09e94c0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# use this to average only the nonzero values. To include also empty values, use np.average(ma)\n",
    "# note this can also be done for 1d array\n",
    "#argument:\n",
    "#ma: matrix of interest\n",
    "\n",
    "#output:\n",
    "# averaged value of matrix\n",
    "def finite_average(ma):\n",
    "    return np.sum(ma)/np.count_nonzero(ma)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53c88b19",
   "metadata": {},
   "source": [
    "# Import data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0ad2af1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get the interval and key information, choose which file to use with file_names\n",
    "\n",
    "interval ,keychr = get_intervals(file_names[0])\n",
    "print(interval)\n",
    "print(keychr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb0d1930",
   "metadata": {},
   "outputs": [],
   "source": [
    "# list of chromosomes\n",
    "chromosomes = [b'chrIV',b'chrXV',b'chrVII',b'chrXII',b'chrXVI',b'chrXIII',b'chrII',b'chrXIV',b'chrX',b'chrXI',b'chrV',b'chrVIII',b'chrIX',b'chrIII',b'chrVI',b'chrI',b'chrM']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb493b66",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get start and end indices corresponding to each chromosomes\n",
    "\n",
    "indices = get_indices(chromosomes,get_end_idx(interval[keychr[0]]))\n",
    "\n",
    "print(indices)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb06a160-19d8-457e-8a8a-426a253c0e60",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import list of genes from rna-seq analysis\n",
    "ds2 = pd.read_csv('ds2h_idx.txt', delimiter='\\t')\n",
    "\n",
    "# indices of of genes from rna-seq analysis\n",
    "chr_idx = ds2.groupby('chromosome_name')['idx'].apply(list).to_dict()\n",
    "\n",
    "#list of chromosomes\n",
    "chr_list = [b'chrI', b'chrII', b'chrIII', b'chrIV', b'chrV', b'chrVI', b'chrVII', b'chrVIII', b'chrIX', b'chrX', b'chrXI', b'chrXII', b'chrXIII', b'chrXIV', b'chrXV', b'chrXVI',b'chrM']\n",
    "\n",
    "chr_idx2 = dict(zip(chr_list, list(chr_idx.values())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "935093ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convinience definations to label and index chromosomes\n",
    "chrome_plots = ['IV','XV','VII','XII','XVI','XIII','II','XIV','X','XI','V','VIII','IX','III','VI','I']\n",
    "chromosomes = [b'chrIV',b'chrXV',b'chrVII',b'chrXII',b'chrXVI',b'chrXIII',b'chrII',b'chrXIV',b'chrX',b'chrXI',b'chrV',b'chrVIII',b'chrIX',b'chrIII',b'chrVI',b'chrI']\n",
    "\n",
    "chromosomes_num= [4, 15, 7, 12, 16, 13, 2, 14, 10, 11, 5, 8, 9, 3, 6, 1]\n",
    "\n",
    "chromosomes_interval = []\n",
    "for i in indices.keys():\n",
    "    if i != b'chrM':\n",
    "        inter = round((indices[i]['start'] + indices[i]['start'])/2)\n",
    "        chromosomes_interval.append(inter)\n",
    "end = indices[b'chrM']['end']\n",
    "\n",
    "chromosome_label = []\n",
    "for i in chromosomes_num:\n",
    "    chromosome_label.append(str(i))\n",
    "# chromosome_label.append('M')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a8ab77c",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "445d8f6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "AA20 = make_symm_matrix(file_names[0])\n",
    "AA22 = make_symm_matrix(file_names[1])\n",
    "AA25 = make_symm_matrix(file_names[2])\n",
    "AA30 = make_symm_matrix(file_names[3])\n",
    "AA32 = make_symm_matrix(file_names[4])\n",
    "AA35 = make_symm_matrix(file_names[5])\n",
    "AA40 = make_symm_matrix(file_names[6])\n",
    "AA42 = make_symm_matrix(file_names[7])\n",
    "AA45 = make_symm_matrix(file_names[8])\n",
    "\n",
    "# AA20 = make_symm_matrix(file_names[5])\n",
    "# AA22 = make_symm_matrix(file_names[3])\n",
    "# AA25 = make_symm_matrix(file_names[6])\n",
    "# AA30 = make_symm_matrix(file_names[4])\n",
    "# AA32 = make_symm_matrix(file_names[2])\n",
    "# AA35 = make_symm_matrix(file_names[7])\n",
    "# AA40 = make_symm_matrix(file_names[0])\n",
    "# AA42 = make_symm_matrix(file_names[1])\n",
    "# AA45 = make_symm_matrix(file_names[8])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6471dbd",
   "metadata": {},
   "source": [
    "# Ratio of different heatmaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c219548b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# log2 ratio of different replicates and time points\n",
    "AA22_AA20 = divide_matrices(AA22,AA20,True)\n",
    "AA25_AA20 = divide_matrices(AA25,AA20,True)\n",
    "\n",
    "AA32_AA30 = divide_matrices(AA32,AA30,True)\n",
    "AA35_AA30 = divide_matrices(AA35,AA30,True)\n",
    "\n",
    "AA42_AA40 =  divide_matrices(AA42,AA40,True)\n",
    "AA45_AA40 = divide_matrices(AA45,AA40,True)\n",
    "\n",
    "AA40_AA30 =  divide_matrices(AA40,AA30,True)\n",
    "AA40_AA20 = divide_matrices(AA40,AA20,True)\n",
    "\n",
    "# ratio of different replicates and time points\n",
    "AA22_AA20_nl = divide_matrices(AA22,AA20,False)\n",
    "AA25_AA20_nl = divide_matrices(AA25,AA20,False)\n",
    "\n",
    "AA32_AA30_nl = divide_matrices(AA32,AA30,False)\n",
    "AA35_AA30_nl = divide_matrices(AA35,AA30,False)\n",
    "\n",
    "AA42_AA40_nl =  divide_matrices(AA42,AA40,False)\n",
    "AA45_AA40_nl = divide_matrices(AA45,AA40,False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0363b65b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.ticker import LogFormatter \n",
    "from matplotlib.ticker import LogFormatterExponent # <-- one new import here\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19117905",
   "metadata": {},
   "source": [
    "------"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2197f197-c34e-4edd-923c-b2e246a4c9a9",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Averaging matrices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10310070",
   "metadata": {},
   "outputs": [],
   "source": [
    "# average heatmaps for 0 hour time point\n",
    "\n",
    "AAave = np.zeros((2441, 2441))\n",
    "for r in np.arange(0, 2441,1):\n",
    "    for c in np.arange(0, 2441, 1):\n",
    "        nom = np.array([ AA20[r, c], AA30[r, c], AA40[r,c]])\n",
    "        denom = np.count_nonzero(nom) #count how many data points in nom are not zero\n",
    "        if denom != 0:\n",
    "            ave = np.sum(nom)/denom\n",
    "            AAave[r, c] = ave\n",
    "# turn anything not finite into 0 in the full matrix        \n",
    "AAave[~np.isfinite(AAave)] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4783b37",
   "metadata": {},
   "outputs": [],
   "source": [
    "# average heatmaps for 2 hour time point\n",
    "\n",
    "AA2ave = np.zeros((2441, 2441))\n",
    "for r in np.arange(0, 2441,1):\n",
    "    for c in np.arange(0, 2441, 1):\n",
    "        nom = np.array([AA22[r, c], AA32[r, c], AA42[r,c]])\n",
    "        denom = np.count_nonzero(nom) #count how many data points in nom are not zero\n",
    "        if denom != 0:\n",
    "            ave = np.sum(nom)/denom\n",
    "            AA2ave[r, c] = ave\n",
    "# turn anything not finite into 0 in the full matrix        \n",
    "AA2ave[~np.isfinite(AA2ave)] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e3900ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# average heatmaps for 5 hour time point\n",
    "\n",
    "AA5ave = np.zeros((2441, 2441))\n",
    "for r in np.arange(0, 2441,1):\n",
    "    for c in np.arange(0, 2441, 1):\n",
    "        nom = np.array([AA25[r, c], AA35[r,c], AA45[r,c]])\n",
    "        denom = np.count_nonzero(nom) #count how many data points in nom are not zero\n",
    "        if denom != 0:\n",
    "            ave = np.sum(nom)/denom\n",
    "            AA5ave[r, c] = ave\n",
    "# turn anything not finite into 0 in the full matrix        \n",
    "AA5ave[~np.isfinite(AA5ave)] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81d6d2e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ratio of 2 hour/DMSO and 5hour/DMSO\n",
    "\n",
    "AAave2_0 = divide_matrices(AA2ave,AAave,log2=True)\n",
    "AAave5_0 = divide_matrices(AA5ave,AAave,log2=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "716b95ed",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Telomere"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8629ce06",
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot settings\n",
    "\n",
    "mpl.rc('xtick', labelsize=12) \n",
    "mpl.rc('ytick', labelsize=12)\n",
    "mpl.rc('figure',figsize = [10.0, 10.0])\n",
    "# matplotlib.rc('axes',titlesize='large')\n",
    "\n",
    "font = {'size'   : 12}\n",
    "\n",
    "mpl.rc('font', **font)\n",
    "\n",
    "from matplotlib.ticker import LogFormatter \n",
    "from matplotlib.ticker import LogFormatterExponent # <-- one new import here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1c83ecb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# obtain an average interaction around a telomere\n",
    "\n",
    "def get_tel(matrix,volcano=3):\n",
    "    volcano_map = np.zeros((2*len(chromosomes),2*len(chromosomes)))\n",
    "\n",
    "    for c in enumerate(chromosomes):\n",
    "        for d in enumerate(chromosomes):\n",
    "            reg = call_region(matrix,c[1],d[1])\n",
    "            reg_a = finite_average(reg[0:volcano,0:volcano])\n",
    "            reg_b = finite_average(reg[0:volcano,-volcano:])\n",
    "    #         reg_b = 0\n",
    "\n",
    "            reg_c = finite_average(reg[-volcano:,0:volcano])\n",
    "    #         reg_c = 0\n",
    "\n",
    "            reg_d = finite_average(reg[-volcano:,-volcano:])\n",
    "            volcano_map[2*c[0],2*d[0]] = reg_a\n",
    "            volcano_map[2*c[0],2*d[0]+1] = reg_b\n",
    "            volcano_map[2*c[0]+1,2*d[0]] = reg_c\n",
    "            volcano_map[2*c[0]+1,2*d[0]+1] = reg_d\n",
    "    return volcano_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3a31fe3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# heatmap of telomere 2hr/DMSO ratio\n",
    "\n",
    "AAave20_tel = get_tel(AAave2_0,volcano=3)\n",
    "\n",
    "testlist=[i.decode('utf-8', 'ignore') for i in chromosomes]\n",
    "\n",
    "chrom_romlist = [testlist[i][3:] for i in range(0,len(testlist))]\n",
    "\n",
    "_=plt.imshow((AAave20_tel),cmap = 'bwr',vmax=2,vmin=-2)\n",
    "_= plt.colorbar(fraction=0.046, pad=0.04)\n",
    "\n",
    "#plt.title('Telomere 2hr/DMSO', fontsize = 20)\n",
    "_=plt.xticks(np.arange(0,32,2),labels = chrom_romlist, rotation = 90)\n",
    "_=plt.yticks(np.arange(0,32,2),labels = chrom_romlist)\n",
    "\n",
    "\n",
    "mpl.rc('xtick', labelsize=20) \n",
    "mpl.rc('ytick', labelsize=20)\n",
    "\n",
    "# plt.savefig('hic234plots/telomere_map_2hr', dpi=200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d021064d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# heatmap of telomere 5hr/DMSO ratio\n",
    "\n",
    "AAave50_tel = get_tel(AAave5_0,volcano=3)\n",
    "\n",
    "_=plt.imshow((AAave50_tel),cmap = 'bwr',vmax=2,vmin=-2)\n",
    "_= plt.colorbar(fraction=0.046, pad=0.04)\n",
    "\n",
    "#plt.title('Telomere 5hr/DMSO')\n",
    "\n",
    "_=plt.xticks(np.arange(0,32,2),labels = chrom_romlist, rotation = 90)\n",
    "_=plt.yticks(np.arange(0,32,2),labels = chrom_romlist)\n",
    "\n",
    "mpl.rc('xtick', labelsize=20) \n",
    "mpl.rc('ytick', labelsize=20)\n",
    "# plt.savefig('hic234plots/telomere_map_5hr', dpi=200)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be23cd5b",
   "metadata": {},
   "source": [
    "-----"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abd3bc74",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Centromere"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b086b739",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get interaction around centromere\n",
    "def get_cent(matrix, volcano=3):\n",
    "    volcano_map = np.zeros((len(chromosomes),len(chromosomes)))\n",
    "    for c in enumerate(chromosomes):\n",
    "        for d in enumerate(chromosomes):\n",
    "            reg = call_region(matrix,c[1],d[1])\n",
    "            cent_index_a = centromere[c[0]]\n",
    "            cent_index_b = centromere[d[0]]\n",
    "            reg_mean = finite_average(reg[cent_index_a-volcano:cent_index_a+volcano+1,cent_index_b-volcano:cent_index_b+volcano+1])\n",
    "            volcano_map[c[0],d[0]]=reg_mean\n",
    "    return volcano_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d00240c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# heatmap of centromeres 2hr/DMSO\n",
    "\n",
    "centromere = [90,65,99,30,111,54,48,126,87,88,30,21,71,23,30,30]\n",
    "\n",
    "AAave20_cent = get_cent(AAave2_0,volcano=1)\n",
    "\n",
    "_=plt.imshow((AAave20_cent),cmap = 'bwr',vmax=2,vmin=-2)\n",
    "_= plt.colorbar(fraction=0.046, pad=0.04)\n",
    "#plt.title('Centromere 2hr/DMSO')\n",
    "\n",
    "mpl.rc('xtick', labelsize=20) \n",
    "mpl.rc('ytick', labelsize=20)\n",
    "\n",
    "_=plt.xticks(np.arange(0,16,1),labels = chrom_romlist, rotation =90)\n",
    "_=plt.yticks(np.arange(0,16,1),labels = chrom_romlist)\n",
    "\n",
    "# plt.savefig('hic234plots/centromere_map_2hr', dpi=200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a13e03a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# heatmap of centromeres 2hr/DMSO\n",
    "\n",
    "AAave50_cent = get_cent(AAave5_0,volcano=1)\n",
    "\n",
    "_=plt.imshow((AAave50_cent),cmap = 'bwr',vmax=2,vmin=-2)\n",
    "_= plt.colorbar(fraction=0.046, pad=0.04)\n",
    "#plt.title('Centromere 5hr/DMSO')\n",
    "mpl.rc('xtick', labelsize=20) \n",
    "mpl.rc('ytick', labelsize=20)\n",
    "\n",
    "_=plt.xticks(np.arange(0,16,1),labels = chrom_romlist, rotation =90)\n",
    "_=plt.yticks(np.arange(0,16,1),labels = chrom_romlist)\n",
    "# plt.savefig('hic234plots/centromere_map_5hr', dpi=200)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c37c906f",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Intrachromosomal Distance Dependence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "672683a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.ndimage import rotate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "960b4507",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function averaging contact frequencies with respect to distance from a gene within chromosome \n",
    "\n",
    "def interchr_length_dependence(matrix,chrome):\n",
    "    reg = call_region(matrix,chrome,chrome)\n",
    "    \n",
    "    m = []\n",
    "    for i in enumerate(reg):\n",
    "        m.append(i[1][i[0]+1:])\n",
    "    df = pd.DataFrame(m)\n",
    "    nd= df.to_numpy()\n",
    "    nd = df.fillna(0).values\n",
    "    nd[nd==0]=np.nan\n",
    "\n",
    "    m = []\n",
    "    for i in enumerate(reg):\n",
    "        a = i[1][0:i[0]]\n",
    "        m.append(np.flip(a))\n",
    "    df2 = pd.DataFrame(m)\n",
    "    nd2 = df2.fillna(0).values\n",
    "    nd2[nd2==0]=np.nan\n",
    "\n",
    "    nd_len = []\n",
    "    nd_len.append(np.nanmean(nd,axis=0))\n",
    "    nd_len.append(np.nanmean(nd2,axis=0))\n",
    "    nd_avg = np.average(np.array(nd_len),axis=0)\n",
    "\n",
    "    return nd_avg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23933739",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sav_gol filter parameters used to smooth the curves\n",
    "\n",
    "win_len=5\n",
    "poly=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12102cad",
   "metadata": {},
   "outputs": [],
   "source": [
    "multiplier = 2\n",
    "mpl.rc('xtick', labelsize=24*multiplier) \n",
    "mpl.rc('ytick', labelsize=24*multiplier)\n",
    "mpl.rc('figure',figsize = [10.0, 10.0])\n",
    "# matplotlib.rc('axes',titlesize='large')\n",
    "\n",
    "font = {'size'   : 24*multiplier}\n",
    "\n",
    "mpl.rc('font', **font)\n",
    "\n",
    "plt.rcParams['axes.titley'] = 0.8    # y is in axes-relative coordinates.\n",
    "plt.rcParams['axes.titlepad'] = -14\n",
    "\n",
    "plt.rcParams['lines.linewidth'] = 4.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37de228a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# rounding function\n",
    "def myround(x, base=50):\n",
    "    return np.int(base * np.round(x/base,1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1c0f0b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#plots for different chromosomes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2547940",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ratio20_list = [AA22_AA20_nl,AA32_AA30_nl,AA42_AA40_nl]\n",
    "ratio50_list = [AA25_AA20_nl,AA35_AA30_nl,AA45_AA40_nl]\n",
    "# b'chrIV',b'chrXV',b'chrVII',b'chrXII',b'chrXVI',b'chrXIII',b'chrII',b'chrXIV',\n",
    "# b'chrX',b'chrXI',b'chrV',b'chrVIII',b'chrIX',b'chrIII',b'chrVI',b'chrI'\n",
    "chromo = b'chrIV'\n",
    "avg_list = []\n",
    "for i in ratio20_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist2 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist2=interaction_dist2[np.isfinite(interaction_dist2)]\n",
    "interaction_dist2 = savgol_filter(interaction_dist2,window_length=win_len,polyorder=poly)\n",
    "\n",
    "avg_list = []\n",
    "for i in ratio50_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist5 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist5=interaction_dist5[np.isfinite(interaction_dist5)]\n",
    "interaction_dist5 = savgol_filter(interaction_dist5,window_length=win_len,polyorder=poly)\n",
    "plt.title(chromo.decode(\"utf-8\"))\n",
    "plt.plot(interaction_dist2, color='green',label='2hr/DMSO')\n",
    "plt.plot(interaction_dist5, color='purple', label='5hr/DMSO')\n",
    "# plt.xticks(np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/2)),np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/2))*5)\n",
    "\n",
    "#copy\n",
    "plt.xticks(np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3)),np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3))*5)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\")[3:],x=0.7, y=0.85)\n",
    " \n",
    "plt.ylim(0.7,2.5)\n",
    "plt.yticks((1,2))\n",
    "plt.vlines(100/5,0.7,2.5,colors='grey',linestyles='--')\n",
    "# plt.ylabel('Fold Change')\n",
    "# plt.xlabel('Distance /kB')\n",
    "# plt.legend()\n",
    "\n",
    "\n",
    "ax1 = plt.subplot()\n",
    "ax1.minorticks_on()\n",
    "ax1.tick_params('both', length=20, width=2, which='major')\n",
    "ax1.tick_params('both', length=10, width=1, which='minor')\n",
    "# plt.savefig('hic234plots/CFvdistance/'+str(chromo.decode(\"utf-8\")), dpi=200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf4e46d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "ratio20_list = [AA22_AA20_nl,AA32_AA30_nl,AA42_AA40_nl]\n",
    "ratio50_list = [AA25_AA20_nl,AA35_AA30_nl,AA45_AA40_nl]\n",
    "# b'chrIV',b'chrXV',b'chrVII',b'chrXII',b'chrXVI',b'chrXIII',b'chrII',b'chrXIV',\n",
    "# b'chrX',b'chrXI',b'chrV',b'chrVIII',b'chrIX',b'chrIII',b'chrVI',b'chrI'\n",
    "chromo = b'chrXV'\n",
    "avg_list = []\n",
    "for i in ratio20_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist2 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist2=interaction_dist2[np.isfinite(interaction_dist2)]\n",
    "interaction_dist2 = savgol_filter(interaction_dist2,window_length=win_len,polyorder=poly)\n",
    "\n",
    "avg_list = []\n",
    "for i in ratio50_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist5 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist5=interaction_dist5[np.isfinite(interaction_dist5)]\n",
    "interaction_dist5 = savgol_filter(interaction_dist5,window_length=win_len,polyorder=poly)\n",
    "\n",
    "plt.plot(interaction_dist2, color='green',label='2hr/DMSO')\n",
    "plt.plot(interaction_dist5, color='purple', label='5hr/DMSO')\n",
    "\n",
    "plt.xticks(np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3)),np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3))*5)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\")[3:],x=0.7, y=0.85)\n",
    " \n",
    "plt.ylim(0.7,2.5)\n",
    "plt.yticks((1,2))\n",
    "plt.vlines(100/5,0.7,2.5,colors='grey',linestyles='--')\n",
    "# plt.ylabel('Fold Change')\n",
    "# plt.xlabel('Distance /kB')\n",
    "# plt.legend()\n",
    "\n",
    "\n",
    "ax1 = plt.subplot()\n",
    "ax1.minorticks_on()\n",
    "ax1.tick_params('both', length=20, width=2, which='major')\n",
    "ax1.tick_params('both', length=10, width=1, which='minor')\n",
    "# plt.savefig('hic234plots/CFvdistance/'+str(chromo.decode(\"utf-8\")), dpi=200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3610a24b",
   "metadata": {},
   "outputs": [],
   "source": [
    "ratio20_list = [AA22_AA20_nl,AA32_AA30_nl,AA42_AA40_nl]\n",
    "ratio50_list = [AA25_AA20_nl,AA35_AA30_nl,AA45_AA40_nl]\n",
    "# b'chrIV',b'chrXV',b'chrVII',b'chrXII',b'chrXVI',b'chrXIII',b'chrII',b'chrXIV',\n",
    "# b'chrX',b'chrXI',b'chrV',b'chrVIII',b'chrIX',b'chrIII',b'chrVI',b'chrI'\n",
    "chromo = b'chrVII'\n",
    "avg_list = []\n",
    "for i in ratio20_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist2 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist2=interaction_dist2[np.isfinite(interaction_dist2)]\n",
    "interaction_dist2 = savgol_filter(interaction_dist2,window_length=win_len,polyorder=poly)\n",
    "\n",
    "avg_list = []\n",
    "for i in ratio50_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist5 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist5=interaction_dist5[np.isfinite(interaction_dist5)]\n",
    "interaction_dist5 = savgol_filter(interaction_dist5,window_length=win_len,polyorder=poly)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\"))\n",
    "plt.plot(interaction_dist2, color='green',label='2hr/DMSO')\n",
    "plt.plot(interaction_dist5, color='purple', label='5hr/DMSO')\n",
    "\n",
    "plt.xticks(np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3)),np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3))*5)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\")[3:],x=0.7, y=0.85)\n",
    " \n",
    "plt.ylim(0.7,2.5)\n",
    "plt.yticks((1,2))\n",
    "plt.vlines(100/5,0.7,2.5,colors='grey',linestyles='--')\n",
    "# plt.ylabel('Fold Change')\n",
    "# plt.xlabel('Distance /kB')\n",
    "# plt.legend()\n",
    "\n",
    "\n",
    "ax1 = plt.subplot()\n",
    "ax1.minorticks_on()\n",
    "ax1.tick_params('both', length=20, width=2, which='major')\n",
    "ax1.tick_params('both', length=10, width=1, which='minor')\n",
    "# plt.savefig('hic234plots/CFvdistance/'+str(chromo.decode(\"utf-8\")), dpi=200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cec44749",
   "metadata": {},
   "outputs": [],
   "source": [
    "ratio20_list = [AA22_AA20_nl,AA32_AA30_nl,AA42_AA40_nl]\n",
    "ratio50_list = [AA25_AA20_nl,AA35_AA30_nl,AA45_AA40_nl]\n",
    "# b'chrIV',b'chrXV',b'chrVII',b'chrXII',b'chrXVI',b'chrXIII',b'chrII',b'chrXIV',\n",
    "# b'chrX',b'chrXI',b'chrV',b'chrVIII',b'chrIX',b'chrIII',b'chrVI',b'chrI'\n",
    "chromo = b'chrXII'\n",
    "avg_list = []\n",
    "for i in ratio20_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist2 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist2=interaction_dist2[np.isfinite(interaction_dist2)]\n",
    "interaction_dist2 = savgol_filter(interaction_dist2,window_length=win_len,polyorder=poly)\n",
    "\n",
    "avg_list = []\n",
    "for i in ratio50_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist5 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist5=interaction_dist5[np.isfinite(interaction_dist5)]\n",
    "interaction_dist5 = savgol_filter(interaction_dist5,window_length=win_len,polyorder=poly)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\"))\n",
    "plt.plot(interaction_dist2, color='green',label='2hr/DMSO')\n",
    "plt.plot(interaction_dist5, color='purple', label='5hr/DMSO')\n",
    "\n",
    "plt.xticks(np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3)),np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3))*5)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\")[3:],x=0.7, y=0.85)\n",
    " \n",
    "plt.ylim(0.7,2.5)\n",
    "plt.yticks((1,2))\n",
    "plt.vlines(100/5,0.7,2.5,colors='grey',linestyles='--')\n",
    "# plt.ylabel('Fold Change')\n",
    "# plt.xlabel('Distance /kB')\n",
    "# plt.legend()\n",
    "\n",
    "\n",
    "ax1 = plt.subplot()\n",
    "ax1.minorticks_on()\n",
    "ax1.tick_params('both', length=20, width=2, which='major')\n",
    "ax1.tick_params('both', length=10, width=1, which='minor')\n",
    "plt.savefig('hic234plots/CFvdistance/'+str(chromo.decode(\"utf-8\")), dpi=200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6080eaca",
   "metadata": {},
   "outputs": [],
   "source": [
    "ratio20_list = [AA22_AA20_nl,AA32_AA30_nl,AA42_AA40_nl]\n",
    "ratio50_list = [AA25_AA20_nl,AA35_AA30_nl,AA45_AA40_nl]\n",
    "# b'chrIV',b'chrXV',b'chrVII',b'chrXII',b'chrXVI',b'chrXIII',b'chrII',b'chrXIV',\n",
    "# b'chrX',b'chrXI',b'chrV',b'chrVIII',b'chrIX',b'chrIII',b'chrVI',b'chrI'\n",
    "chromo = b'chrXVI'\n",
    "avg_list = []\n",
    "for i in ratio20_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist2 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist2=interaction_dist2[np.isfinite(interaction_dist2)]\n",
    "interaction_dist2 = savgol_filter(interaction_dist2,window_length=win_len,polyorder=poly)\n",
    "\n",
    "avg_list = []\n",
    "for i in ratio50_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist5 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist5=interaction_dist5[np.isfinite(interaction_dist5)]\n",
    "interaction_dist5 = savgol_filter(interaction_dist5,window_length=win_len,polyorder=poly)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\"))\n",
    "plt.plot(interaction_dist2, color='green',label='2hr/DMSO')\n",
    "plt.plot(interaction_dist5, color='purple', label='5hr/DMSO')\n",
    "\n",
    "plt.xticks(np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3)),np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3))*5)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\")[3:],x=0.7, y=0.85)\n",
    " \n",
    "plt.ylim(0.7,2.5)\n",
    "plt.yticks((1,2))\n",
    "plt.vlines(100/5,0.7,2.5,colors='grey',linestyles='--')\n",
    "# plt.ylabel('Fold Change')\n",
    "# plt.xlabel('Distance /kB')\n",
    "# plt.legend()\n",
    "\n",
    "\n",
    "ax1 = plt.subplot()\n",
    "ax1.minorticks_on()\n",
    "ax1.tick_params('both', length=20, width=2, which='major')\n",
    "ax1.tick_params('both', length=10, width=1, which='minor')\n",
    "plt.savefig('hic234plots/CFvdistance/'+str(chromo.decode(\"utf-8\")), dpi=200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbd9a5e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "ratio20_list = [AA22_AA20_nl,AA32_AA30_nl,AA42_AA40_nl]\n",
    "ratio50_list = [AA25_AA20_nl,AA35_AA30_nl,AA45_AA40_nl]\n",
    "# b'chrIV',b'chrXV',b'chrVII',b'chrXII',b'chrXVI',b'chrXIII',b'chrII',b'chrXIV',\n",
    "# b'chrX',b'chrXI',b'chrV',b'chrVIII',b'chrIX',b'chrIII',b'chrVI',b'chrI'\n",
    "chromo = b'chrXIII'\n",
    "avg_list = []\n",
    "for i in ratio20_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist2 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist2=interaction_dist2[np.isfinite(interaction_dist2)]\n",
    "interaction_dist2 = savgol_filter(interaction_dist2,window_length=win_len,polyorder=poly)\n",
    "\n",
    "avg_list = []\n",
    "for i in ratio50_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist5 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist5=interaction_dist5[np.isfinite(interaction_dist5)]\n",
    "interaction_dist5 = savgol_filter(interaction_dist5,window_length=win_len,polyorder=poly)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\"))\n",
    "plt.plot(interaction_dist2, color='green',label='2hr/DMSO')\n",
    "plt.plot(interaction_dist5, color='purple', label='5hr/DMSO')\n",
    "\n",
    "plt.xticks(np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3)),np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3))*5)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\")[3:],x=0.7, y=0.85)\n",
    " \n",
    "plt.ylim(0.7,2.5)\n",
    "plt.yticks((1,2))\n",
    "plt.vlines(100/5,0.7,2.5,colors='grey',linestyles='--')\n",
    "# plt.ylabel('Fold Change')\n",
    "# plt.xlabel('Distance /kB')\n",
    "# plt.legend()\n",
    "\n",
    "\n",
    "ax1 = plt.subplot()\n",
    "ax1.minorticks_on()\n",
    "ax1.tick_params('both', length=20, width=2, which='major')\n",
    "ax1.tick_params('both', length=10, width=1, which='minor')\n",
    "plt.savefig('hic234plots/CFvdistance/'+str(chromo.decode(\"utf-8\")), dpi=200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18689484",
   "metadata": {},
   "outputs": [],
   "source": [
    "ratio20_list = [AA22_AA20_nl,AA32_AA30_nl,AA42_AA40_nl]\n",
    "ratio50_list = [AA25_AA20_nl,AA35_AA30_nl,AA45_AA40_nl]\n",
    "# b'chrIV',b'chrXV',b'chrVII',b'chrXII',b'chrXVI',b'chrXIII',b'chrII',b'chrXIV',\n",
    "# b'chrX',b'chrXI',b'chrV',b'chrVIII',b'chrIX',b'chrIII',b'chrVI',b'chrI'\n",
    "chromo = b'chrII'\n",
    "avg_list = []\n",
    "for i in ratio20_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist2 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist2=interaction_dist2[np.isfinite(interaction_dist2)]\n",
    "interaction_dist2 = savgol_filter(interaction_dist2,window_length=win_len,polyorder=poly)\n",
    "\n",
    "avg_list = []\n",
    "for i in ratio50_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist5 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist5=interaction_dist5[np.isfinite(interaction_dist5)]\n",
    "interaction_dist5 = savgol_filter(interaction_dist5,window_length=win_len,polyorder=poly)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\"))\n",
    "plt.plot(interaction_dist2, color='green',label='2hr/DMSO')\n",
    "plt.plot(interaction_dist5, color='purple', label='5hr/DMSO')\n",
    "\n",
    "plt.xticks(np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3)),np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3))*5)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\")[3:],x=0.7, y=0.85)\n",
    " \n",
    "plt.ylim(0.7,2.5)\n",
    "plt.yticks((1,2))\n",
    "plt.vlines(100/5,0.7,2.5,colors='grey',linestyles='--')\n",
    "# plt.ylabel('Fold Change')\n",
    "# plt.xlabel('Distance /kB')\n",
    "# plt.legend()\n",
    "\n",
    "\n",
    "ax1 = plt.subplot()\n",
    "ax1.minorticks_on()\n",
    "ax1.tick_params('both', length=20, width=2, which='major')\n",
    "ax1.tick_params('both', length=10, width=1, which='minor')\n",
    "plt.savefig('hic234plots/CFvdistance/'+str(chromo.decode(\"utf-8\")), dpi=200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a7d7e68",
   "metadata": {},
   "outputs": [],
   "source": [
    "ratio20_list = [AA22_AA20_nl,AA32_AA30_nl,AA42_AA40_nl]\n",
    "ratio50_list = [AA25_AA20_nl,AA35_AA30_nl,AA45_AA40_nl]\n",
    "# b'chrIV',b'chrXV',b'chrVII',b'chrXII',b'chrXVI',b'chrXIII',b'chrII',b'chrXIV',\n",
    "# b'chrX',b'chrXI',b'chrV',b'chrVIII',b'chrIX',b'chrIII',b'chrVI',b'chrI'\n",
    "chromo = b'chrXIV'\n",
    "avg_list = []\n",
    "for i in ratio20_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist2 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist2=interaction_dist2[np.isfinite(interaction_dist2)]\n",
    "interaction_dist2 = savgol_filter(interaction_dist2,window_length=win_len,polyorder=poly)\n",
    "\n",
    "avg_list = []\n",
    "for i in ratio50_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist5 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist5=interaction_dist5[np.isfinite(interaction_dist5)]\n",
    "interaction_dist5 = savgol_filter(interaction_dist5,window_length=win_len,polyorder=poly)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\"))\n",
    "plt.plot(interaction_dist2, color='green',label='2hr/DMSO')\n",
    "plt.plot(interaction_dist5, color='purple', label='5hr/DMSO')\n",
    "\n",
    "plt.xticks(np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3)),np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3))*5)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\")[3:],x=0.7, y=0.85)\n",
    " \n",
    "plt.ylim(0.7,2.5)\n",
    "plt.yticks((1,2))\n",
    "plt.vlines(100/5,0.7,2.5,colors='grey',linestyles='--')\n",
    "# plt.ylabel('Fold Change')\n",
    "# plt.xlabel('Distance /kB')\n",
    "# plt.legend()\n",
    "\n",
    "\n",
    "ax1 = plt.subplot()\n",
    "ax1.minorticks_on()\n",
    "ax1.tick_params('both', length=20, width=2, which='major')\n",
    "ax1.tick_params('both', length=10, width=1, which='minor')\n",
    "plt.savefig('hic234plots/CFvdistance/'+str(chromo.decode(\"utf-8\")), dpi=200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4edee510",
   "metadata": {},
   "outputs": [],
   "source": [
    "ratio20_list = [AA22_AA20_nl,AA32_AA30_nl,AA42_AA40_nl]\n",
    "ratio50_list = [AA25_AA20_nl,AA35_AA30_nl,AA45_AA40_nl]\n",
    "# b'chrIV',b'chrXV',b'chrVII',b'chrXII',b'chrXVI',b'chrXIII',b'chrII',b'chrXIV',\n",
    "# b'chrX',b'chrXI',b'chrV',b'chrVIII',b'chrIX',b'chrIII',b'chrVI',b'chrI'\n",
    "chromo = b'chrX'\n",
    "avg_list = []\n",
    "for i in ratio20_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist2 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist2=interaction_dist2[np.isfinite(interaction_dist2)]\n",
    "interaction_dist2 = savgol_filter(interaction_dist2,window_length=win_len,polyorder=poly)\n",
    "\n",
    "avg_list = []\n",
    "for i in ratio50_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist5 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist5=interaction_dist5[np.isfinite(interaction_dist5)]\n",
    "interaction_dist5 = savgol_filter(interaction_dist5,window_length=win_len,polyorder=poly)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\"))\n",
    "plt.plot(interaction_dist2, color='green',label='2hr/DMSO')\n",
    "plt.plot(interaction_dist5, color='purple', label='5hr/DMSO')\n",
    "\n",
    "plt.xticks(np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3)),np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3))*5)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\")[3:],x=0.7, y=0.85)\n",
    " \n",
    "plt.ylim(0.7,2.5)\n",
    "plt.yticks((1,2))\n",
    "plt.vlines(100/5,0.7,2.5,colors='grey',linestyles='--')\n",
    "# plt.ylabel('Fold Change')\n",
    "# plt.xlabel('Distance /kB')\n",
    "# plt.legend()\n",
    "\n",
    "\n",
    "ax1 = plt.subplot()\n",
    "ax1.minorticks_on()\n",
    "ax1.tick_params('both', length=20, width=2, which='major')\n",
    "ax1.tick_params('both', length=10, width=1, which='minor')\n",
    "plt.savefig('hic234plots/CFvdistance/'+str(chromo.decode(\"utf-8\")), dpi=200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67b3a5be",
   "metadata": {},
   "outputs": [],
   "source": [
    "ratio20_list = [AA22_AA20_nl,AA32_AA30_nl,AA42_AA40_nl]\n",
    "ratio50_list = [AA25_AA20_nl,AA35_AA30_nl,AA45_AA40_nl]\n",
    "# b'chrIV',b'chrXV',b'chrVII',b'chrXII',b'chrXVI',b'chrXIII',b'chrII',b'chrXIV',\n",
    "# b'chrX',b'chrXI',b'chrV',b'chrVIII',b'chrIX',b'chrIII',b'chrVI',b'chrI'\n",
    "chromo = b'chrXI'\n",
    "avg_list = []\n",
    "for i in ratio20_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist2 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist2=interaction_dist2[np.isfinite(interaction_dist2)]\n",
    "interaction_dist2 = savgol_filter(interaction_dist2,window_length=win_len,polyorder=poly)\n",
    "\n",
    "avg_list = []\n",
    "for i in ratio50_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist5 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist5=interaction_dist5[np.isfinite(interaction_dist5)]\n",
    "interaction_dist5 = savgol_filter(interaction_dist5,window_length=win_len,polyorder=poly)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\"))\n",
    "plt.plot(interaction_dist2, color='green',label='2hr/DMSO')\n",
    "plt.plot(interaction_dist5, color='purple', label='5hr/DMSO')\n",
    "\n",
    "plt.xticks(np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3)),np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3))*5)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\")[3:],x=0.7, y=0.85)\n",
    " \n",
    "plt.ylim(0.7,2.5)\n",
    "plt.yticks((1,2))\n",
    "plt.vlines(100/5,0.7,2.5,colors='grey',linestyles='--')\n",
    "# plt.ylabel('Fold Change')\n",
    "# plt.xlabel('Distance /kB')\n",
    "# plt.legend()\n",
    "\n",
    "\n",
    "ax1 = plt.subplot()\n",
    "ax1.minorticks_on()\n",
    "ax1.tick_params('both', length=20, width=2, which='major')\n",
    "ax1.tick_params('both', length=10, width=1, which='minor')\n",
    "plt.savefig('hic234plots/CFvdistance/'+str(chromo.decode(\"utf-8\")), dpi=200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86a33b9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ratio20_list = [AA22_AA20_nl,AA32_AA30_nl,AA42_AA40_nl]\n",
    "ratio50_list = [AA25_AA20_nl,AA35_AA30_nl,AA45_AA40_nl]\n",
    "# b'chrIV',b'chrXV',b'chrVII',b'chrXII',b'chrXVI',b'chrXIII',b'chrII',b'chrXIV',\n",
    "# b'chrX',b'chrXI',b'chrV',b'chrVIII',b'chrIX',b'chrIII',b'chrVI',b'chrI'\n",
    "chromo = b'chrV'\n",
    "avg_list = []\n",
    "for i in ratio20_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist2 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist2=interaction_dist2[np.isfinite(interaction_dist2)]\n",
    "interaction_dist2 = savgol_filter(interaction_dist2,window_length=win_len,polyorder=poly)\n",
    "\n",
    "avg_list = []\n",
    "for i in ratio50_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist5 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist5=interaction_dist5[np.isfinite(interaction_dist5)]\n",
    "interaction_dist5 = savgol_filter(interaction_dist5,window_length=win_len,polyorder=poly)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\"))\n",
    "plt.plot(interaction_dist2, color='green',label='2hr/DMSO')\n",
    "plt.plot(interaction_dist5, color='purple', label='5hr/DMSO')\n",
    "\n",
    "plt.xticks(np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3.05)),np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3.05))*5)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\")[3:],x=0.7, y=0.85)\n",
    " \n",
    "plt.ylim(0.7,2.5)\n",
    "plt.yticks((1,2))\n",
    "plt.vlines(100/5,0.7,2.5,colors='grey',linestyles='--')\n",
    "# plt.ylabel('Fold Change')\n",
    "# plt.xlabel('Distance /kB')\n",
    "# plt.legend()\n",
    "\n",
    "\n",
    "ax1 = plt.subplot()\n",
    "ax1.minorticks_on()\n",
    "ax1.tick_params('both', length=20, width=2, which='major')\n",
    "ax1.tick_params('both', length=10, width=1, which='minor')\n",
    "plt.savefig('hic234plots/CFvdistance/'+str(chromo.decode(\"utf-8\")), dpi=200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a991d63",
   "metadata": {},
   "outputs": [],
   "source": [
    "ratio20_list = [AA22_AA20_nl,AA32_AA30_nl,AA42_AA40_nl]\n",
    "ratio50_list = [AA25_AA20_nl,AA35_AA30_nl,AA45_AA40_nl]\n",
    "# b'chrIV',b'chrXV',b'chrVII',b'chrXII',b'chrXVI',b'chrXIII',b'chrII',b'chrXIV',\n",
    "# b'chrX',b'chrXI',b'chrV',b'chrVIII',b'chrIX',b'chrIII',b'chrVI',b'chrI'\n",
    "chromo = b'chrVIII'\n",
    "avg_list = []\n",
    "for i in ratio20_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist2 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist2=interaction_dist2[np.isfinite(interaction_dist2)]\n",
    "interaction_dist2 = savgol_filter(interaction_dist2,window_length=win_len,polyorder=poly)\n",
    "\n",
    "avg_list = []\n",
    "for i in ratio50_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist5 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist5=interaction_dist5[np.isfinite(interaction_dist5)]\n",
    "interaction_dist5 = savgol_filter(interaction_dist5,window_length=win_len,polyorder=poly)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\"))\n",
    "plt.plot(interaction_dist2, color='green',label='2hr/DMSO')\n",
    "plt.plot(interaction_dist5, color='purple', label='5hr/DMSO')\n",
    "\n",
    "plt.xticks(np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3.05)),np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3.05))*5)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\")[3:],x=0.7, y=0.85)\n",
    " \n",
    "plt.ylim(0.7,2.5)\n",
    "plt.yticks((1,2))\n",
    "plt.vlines(100/5,0.7,2.5,colors='grey',linestyles='--')\n",
    "# plt.ylabel('Fold Change')\n",
    "# plt.xlabel('Distance /kB')\n",
    "# plt.legend()\n",
    "\n",
    "\n",
    "ax1 = plt.subplot()\n",
    "ax1.minorticks_on()\n",
    "ax1.tick_params('both', length=20, width=2, which='major')\n",
    "ax1.tick_params('both', length=10, width=1, which='minor')\n",
    "plt.savefig('hic234plots/CFvdistance/'+str(chromo.decode(\"utf-8\")), dpi=200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ce0b593",
   "metadata": {},
   "outputs": [],
   "source": [
    "ratio20_list = [AA22_AA20_nl,AA32_AA30_nl,AA42_AA40_nl]\n",
    "ratio50_list = [AA25_AA20_nl,AA35_AA30_nl,AA45_AA40_nl]\n",
    "# b'chrIV',b'chrXV',b'chrVII',b'chrXII',b'chrXVI',b'chrXIII',b'chrII',b'chrXIV',\n",
    "# b'chrX',b'chrXI',b'chrV',b'chrVIII',b'chrIX',b'chrIII',b'chrVI',b'chrI'\n",
    "chromo = b'chrIX'\n",
    "avg_list = []\n",
    "for i in ratio20_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist2 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist2=interaction_dist2[np.isfinite(interaction_dist2)]\n",
    "interaction_dist2 = savgol_filter(interaction_dist2,window_length=win_len,polyorder=poly)\n",
    "\n",
    "avg_list = []\n",
    "for i in ratio50_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist5 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist5=interaction_dist5[np.isfinite(interaction_dist5)]\n",
    "interaction_dist5 = savgol_filter(interaction_dist5,window_length=win_len,polyorder=poly)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\"))\n",
    "plt.plot(interaction_dist2, color='green',label='2hr/DMSO')\n",
    "plt.plot(interaction_dist5, color='purple', label='5hr/DMSO')\n",
    "\n",
    "plt.xticks(np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3)),np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3))*5)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\")[3:],x=0.7, y=0.85)\n",
    " \n",
    "plt.ylim(0.7,2.5)\n",
    "plt.yticks((1,2))\n",
    "plt.vlines(100/5,0.7,2.5,colors='grey',linestyles='--')\n",
    "# plt.ylabel('Fold Change')\n",
    "# plt.xlabel('Distance /kB')\n",
    "# plt.legend()\n",
    "\n",
    "\n",
    "ax1 = plt.subplot()\n",
    "ax1.minorticks_on()\n",
    "ax1.tick_params('both', length=20, width=2, which='major')\n",
    "ax1.tick_params('both', length=10, width=1, which='minor')\n",
    "plt.savefig('hic234plots/CFvdistance/'+str(chromo.decode(\"utf-8\")), dpi=200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edf541df",
   "metadata": {},
   "outputs": [],
   "source": [
    "ratio20_list = [AA22_AA20_nl,AA32_AA30_nl,AA42_AA40_nl]\n",
    "ratio50_list = [AA25_AA20_nl,AA35_AA30_nl,AA45_AA40_nl]\n",
    "# b'chrIV',b'chrXV',b'chrVII',b'chrXII',b'chrXVI',b'chrXIII',b'chrII',b'chrXIV',\n",
    "# b'chrX',b'chrXI',b'chrV',b'chrVIII',b'chrIX',b'chrIII',b'chrVI',b'chrI'\n",
    "chromo = b'chrIII'\n",
    "avg_list = []\n",
    "for i in ratio20_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist2 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist2=interaction_dist2[np.isfinite(interaction_dist2)]\n",
    "interaction_dist2 = savgol_filter(interaction_dist2,window_length=win_len,polyorder=poly)\n",
    "\n",
    "avg_list = []\n",
    "for i in ratio50_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist5 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist5=interaction_dist5[np.isfinite(interaction_dist5)]\n",
    "interaction_dist5 = savgol_filter(interaction_dist5,window_length=win_len,polyorder=poly)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\"))\n",
    "plt.plot(interaction_dist2, color='green',label='2hr/DMSO')\n",
    "plt.plot(interaction_dist5, color='purple', label='5hr/DMSO')\n",
    "\n",
    "plt.xticks(np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3)),np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3))*5)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\")[3:],x=0.7, y=0.85)\n",
    " \n",
    "plt.ylim(0.7,2.5)\n",
    "plt.yticks((1,2))\n",
    "plt.vlines(100/5,0.7,2.5,colors='grey',linestyles='--')\n",
    "# plt.ylabel('Fold Change')\n",
    "# plt.xlabel('Distance /kB')\n",
    "# plt.legend()\n",
    "\n",
    "\n",
    "ax1 = plt.subplot()\n",
    "ax1.minorticks_on()\n",
    "ax1.tick_params('both', length=20, width=2, which='major')\n",
    "ax1.tick_params('both', length=10, width=1, which='minor')\n",
    "plt.savefig('hic234plots/CFvdistance/'+str(chromo.decode(\"utf-8\")), dpi=200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17f4dabd",
   "metadata": {},
   "outputs": [],
   "source": [
    "ratio20_list = [AA22_AA20_nl,AA32_AA30_nl,AA42_AA40_nl]\n",
    "ratio50_list = [AA25_AA20_nl,AA35_AA30_nl,AA45_AA40_nl]\n",
    "# b'chrIV',b'chrXV',b'chrVII',b'chrXII',b'chrXVI',b'chrXIII',b'chrII',b'chrXIV',\n",
    "# b'chrX',b'chrXI',b'chrV',b'chrVIII',b'chrIX',b'chrIII',b'chrVI',b'chrI'\n",
    "chromo = b'chrVI'\n",
    "avg_list = []\n",
    "for i in ratio20_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist2 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist2=interaction_dist2[np.isfinite(interaction_dist2)]\n",
    "interaction_dist2 = savgol_filter(interaction_dist2,window_length=win_len,polyorder=poly)\n",
    "\n",
    "avg_list = []\n",
    "for i in ratio50_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist5 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist5=interaction_dist5[np.isfinite(interaction_dist5)]\n",
    "interaction_dist5 = savgol_filter(interaction_dist5,window_length=win_len,polyorder=poly)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\"))\n",
    "plt.plot(interaction_dist2, color='green',label='2hr/DMSO')\n",
    "plt.plot(interaction_dist5, color='purple', label='5hr/DMSO')\n",
    "\n",
    "plt.xticks(np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3.01)),np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3.01))*5)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\")[3:],x=0.7, y=0.85)\n",
    " \n",
    "plt.ylim(0.7,2.5)\n",
    "plt.yticks((1,2))\n",
    "plt.vlines(100/5,0.7,2.5,colors='grey',linestyles='--')\n",
    "# plt.ylabel('Fold Change')\n",
    "# plt.xlabel('Distance /kB')\n",
    "# plt.legend()\n",
    "\n",
    "\n",
    "ax1 = plt.subplot()\n",
    "ax1.minorticks_on()\n",
    "ax1.tick_params('both', length=20, width=2, which='major')\n",
    "ax1.tick_params('both', length=10, width=1, which='minor')\n",
    "plt.savefig('hic234plots/CFvdistance/'+str(chromo.decode(\"utf-8\")), dpi=200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37bb2ffa",
   "metadata": {},
   "outputs": [],
   "source": [
    "ratio20_list = [AA22_AA20_nl,AA32_AA30_nl,AA42_AA40_nl]\n",
    "ratio50_list = [AA25_AA20_nl,AA35_AA30_nl,AA45_AA40_nl]\n",
    "# b'chrIV',b'chrXV',b'chrVII',b'chrXII',b'chrXVI',b'chrXIII',b'chrII',b'chrXIV',\n",
    "# b'chrX',b'chrXI',b'chrV',b'chrVIII',b'chrIX',b'chrIII',b'chrVI',b'chrI'\n",
    "chromo = b'chrI'\n",
    "avg_list = []\n",
    "for i in ratio20_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist2 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist2=interaction_dist2[np.isfinite(interaction_dist2)]\n",
    "interaction_dist2 = savgol_filter(interaction_dist2,window_length=win_len,polyorder=poly)\n",
    "\n",
    "avg_list = []\n",
    "for i in ratio50_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist5 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist5=interaction_dist5[np.isfinite(interaction_dist5)]\n",
    "interaction_dist5 = savgol_filter(interaction_dist5,window_length=win_len,polyorder=poly)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\"))\n",
    "plt.plot(interaction_dist2, color='green',label='2hr/DMSO')\n",
    "plt.plot(interaction_dist5, color='purple', label='5hr/DMSO')\n",
    "\n",
    "plt.xticks(np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3)),np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/3))*5)\n",
    "\n",
    "plt.title(chromo.decode(\"utf-8\")[3:],x=0.7, y=0.85)\n",
    " \n",
    "plt.ylim(0.7,2.5)\n",
    "plt.yticks((1,2))\n",
    "plt.vlines(100/5,0.7,2.5,colors='grey',linestyles='--')\n",
    "# plt.ylabel('Fold Change')\n",
    "# plt.xlabel('Distance /kB')\n",
    "# plt.legend()\n",
    "\n",
    "\n",
    "ax1 = plt.subplot()\n",
    "ax1.minorticks_on()\n",
    "ax1.tick_params('both', length=20, width=2, which='major')\n",
    "ax1.tick_params('both', length=10, width=1, which='minor')\n",
    "plt.savefig('hic234plots/CFvdistance/'+str(chromo.decode(\"utf-8\")), dpi=200)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1dc52019-4eff-485e-872b-7434f461f747",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Others"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18edd614-4f5b-4d14-bcfa-2e3e10540531",
   "metadata": {},
   "outputs": [],
   "source": [
    "multiplier = 1\n",
    "mpl.rc('xtick', labelsize=24*multiplier) \n",
    "mpl.rc('ytick', labelsize=24*multiplier)\n",
    "mpl.rc('figure',figsize = [10.0, 10.0])\n",
    "# matplotlib.rc('axes',titlesize='large')\n",
    "\n",
    "font = {'size'   : 24*multiplier}\n",
    "\n",
    "mpl.rc('font', **font)\n",
    "\n",
    "plt.rcParams['axes.titley'] = 0.8    # y is in axes-relative coordinates.\n",
    "plt.rcParams['axes.titlepad'] = -14\n",
    "\n",
    "plt.rcParams['lines.linewidth'] = 2.5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "685bab4a-37a3-428e-ae03-994b1774bb1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ratio20_list = [AA22_AA20_nl,AA32_AA30_nl,AA42_AA40_nl]\n",
    "ratio50_list = [AA25_AA20_nl,AA35_AA30_nl,AA45_AA40_nl]\n",
    "# b'chrIV',b'chrXV',b'chrVII',b'chrXII',b'chrXVI',b'chrXIII',b'chrII',b'chrXIV',\n",
    "# b'chrX',b'chrXI',b'chrV',b'chrVIII',b'chrIX',b'chrIII',b'chrVI',b'chrI'\n",
    "chromo = b'chrII'\n",
    "avg_list = []\n",
    "for i in ratio20_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist2 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist2=interaction_dist2[np.isfinite(interaction_dist2)]\n",
    "interaction_dist2 = savgol_filter(interaction_dist2,window_length=win_len,polyorder=poly)\n",
    "\n",
    "avg_list = []\n",
    "for i in ratio50_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist5 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist5=interaction_dist5[np.isfinite(interaction_dist5)]\n",
    "interaction_dist5 = savgol_filter(interaction_dist5,window_length=win_len,polyorder=poly)\n",
    "\n",
    "# plt.title(chromo.decode(\"utf-8\"))\n",
    "plt.plot(interaction_dist2, color='green',label='2 hours')\n",
    "plt.plot(interaction_dist5, color='purple', label='5 hours')\n",
    "plt.xticks(np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/8)),np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/8))*5)\n",
    "plt.ylim(0.75,2.3)\n",
    "plt.vlines(100/5,0.8,2.2,colors='grey',linestyles='--')\n",
    "plt.ylabel('Fold Change')\n",
    "plt.xlabel('Distance /kB')\n",
    "plt.legend(fontsize=\"30\", loc =\"upper right\")\n",
    "plt.savefig('hic234plots/CFvdistance/other/'+str(chromo.decode(\"utf-8\")), dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98167646-f476-4ae2-93d9-d8511614d2b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "ratio20_list = [AA22_AA20_nl,AA32_AA30_nl,AA42_AA40_nl]\n",
    "ratio50_list = [AA25_AA20_nl,AA35_AA30_nl,AA45_AA40_nl]\n",
    "# b'chrIV',b'chrXV',b'chrVII',b'chrXII',b'chrXVI',b'chrXIII',b'chrII',b'chrXIV',\n",
    "# b'chrX',b'chrXI',b'chrV',b'chrVIII',b'chrIX',b'chrIII',b'chrVI',b'chrI'\n",
    "chromo = b'chrIV'\n",
    "avg_list = []\n",
    "for i in ratio20_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist2 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist2=interaction_dist2[np.isfinite(interaction_dist2)]\n",
    "interaction_dist2 = savgol_filter(interaction_dist2,window_length=win_len,polyorder=poly)\n",
    "\n",
    "avg_list = []\n",
    "for i in ratio50_list:\n",
    "    avg_list.append(interchr_length_dependence(i,chromo))\n",
    "interaction_dist5 = np.nanmean(np.array(avg_list),axis=0)\n",
    "interaction_dist5=interaction_dist5[np.isfinite(interaction_dist5)]\n",
    "interaction_dist5 = savgol_filter(interaction_dist5,window_length=win_len,polyorder=poly)\n",
    "\n",
    "# plt.title(chromo.decode(\"utf-8\"))\n",
    "plt.plot(interaction_dist2, color='green',label='2 hours')\n",
    "plt.plot(interaction_dist5, color='purple', label='5 hours')\n",
    "plt.xticks(np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/5)),np.arange(len(interaction_dist2),step=int(len(interaction_dist2)/5))*5)\n",
    "plt.ylim(0.775,2.25)\n",
    "\n",
    "plt.vlines(25/5,0.8,2.2,colors='grey',linestyles='--')\n",
    "plt.vlines(100/5,0.8,2.2,colors='grey',linestyles='--')\n",
    "plt.vlines(500/5,0.8,2.2,colors='grey',linestyles='--')\n",
    "\n",
    "plt.ylabel('Fold Change')\n",
    "plt.xlabel('Distance /kB')\n",
    "plt.legend(fontsize=\"30\", loc =\"upper right\")\n",
    "plt.savefig('hic234plots/CFvdistance/other/'+str(chromo.decode(\"utf-8\")), dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37b63355-27cb-4623-9f3c-97966ef78fc6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
